### PR TITLE
Ignore non-causal mask in more cases with SDPA

### LINKED ISSUE #30095: `_prepare_4d_attention_mask_for_sdpa` is not for causal attention but claims...
... SDPA causal mask generation may be wrong for the mask generation.

https://github.com/huggingface/transformers/blob/76fa17c1663a0efeca7208c20579833365584889/src/transformers/modeling_attn_mask_utils.py#L421-L433

Will it be safe to just return `None` for the `else:` case?

For causal attention, we can just use `_prepare_4d_causal_attention_mask_for_sdpa`

Related issues:
https://github.com/pytorch/pytorch/issues/108108
https://github.com/Dao-AILab/flash-attention/commit/9e5e8bc91e30af5cdc321362b553f6c0da332e30
https://github.com/huggingface/transformers/pull/28802

### PR DESCRIPTION
**Pull request by fxmarty**
Fixes https://github.com/huggingface/transformers/issues/30095. In the non-causal case, we can ignore the mask even for `key_value_length != tgt_len` as we will never have any fully masked row and will never hit the SDPA's mem-efficient attention backend issue.

Depends on https://github.com/huggingface/transformers/pull/28802

### CODE REVIEW

File: src/transformers/modeling_attn_mask_utils.py [hunk 1]
```diff
@@ -406,31 +406,18 @@ def _prepare_4d_attention_mask_for_sdpa(mask: torch.Tensor, dtype: torch.dtype,
         tgt_len (`int`):
             The target length or query length the created mask shall have.
     """
-    batch_size, key_value_length = mask.shape
+    _, key_value_length = mask.shape
```
Comment (minostauros) [inline src/transformers/modeling_attn_mask_utils.py:409]: Can we change the input arg `mask: torch.Tensor` to `mask: Optional[torch.Tensor]` and `return None` immediately if `mask is None`? The docstring is not compliant with the actual input. (`mask (`torch.Tensor` or `None`):`)
Will it break the `is_tracing` check?

Comment (fxmarty) [reply src/transformers/modeling_attn_mask_utils.py:409]: @minostauros Yes indeed ideally we would want to do that. In practice, the calls to these functions in modeling files are always guarded by:
```python
if attention_mask is not None:
```

but we should IMO indeed accept `Optional[torch.Tensor]`. I'll leave that to an other PR.

**Review by ArthurZucker — APPROVED** [review]
Good cleanup IMO. Let's make sure the workflow is triggered for this

**Comment by ArthurZucker** [comment]
This is probably fixed on main
